book/src/dev/rfcs/0004-asynchronous-script-verification.md

```diff
@@ -175,23 +360,33 @@ impl PendingUtxos {
 
 The state service should maintain an `Arc<Mutex<PendingUtxos>>`, used as follows:
 
-1. In `Service::call(Request::AwaitUtxo(u))`, the service should:
+1. In `Service::call(Request::AwaitSpendableUtxo { outpoint: u, .. }`, the service should:
   - call `PendingUtxos::queue(u)` to get a future `f` to return to the caller;
-  spawn a task that does a rocksdb lookup for `u`, calling `PendingUtxos::respond(u, output)` if present;
+  - spawn a task that does a rocksdb lookup for `u`, calling `PendingUtxos::respond(u, output)` if present;
   - check the in-memory storage for `u`, calling `PendingUtxos::respond(u, output)` if present;
   - return `f` to the caller (it may already be ready).
-  The common case is that `u` references an old UTXO, so spawning the lookup
+  The common case is that `u` references an old spendable UTXO, so spawning the lookup
   task first means that we don't wait to check in-memory storage for `u`
   before starting the rocksdb lookup.
 
-2. In `Service::call(Request::CommitBlock(block, ..))`, the service should:
-  - call `PendingUtxos::check_block(block.as_ref())`;
+2. In `f`, the future returned by `PendingUtxos::queue(u)`, the service should
+  check that the `Utxo` is spendable before returning it:
+  - if `Utxo.from_coinbase` is false, return the utxo;
+  - if `Utxo.from_coinbase` is true, check that:
+    - `spend_restriction` is `AllShieldedOutputs`, and
+    - `spend_height` is greater than or equal to
+      `MIN_TRANSPARENT_COINBASE_MATURITY` plus the `Utxo.height`,
+    - then return the utxo.
```

Comment:
Optional nit: The last line seemed a bit out of place, so I tried to improve the structure of the text. I'm not sure if it helps, or if it risks making things ambiguous, so feel free to ignore this.

```suggestion
  - if `Utxo.from_coinbase` is true, only return the utxo if:
    - `spend_restriction` is `AllShieldedOutputs`, and
    - `spend_height` is greater than or equal to
      `MIN_TRANSPARENT_COINBASE_MATURITY` plus the `Utxo.height`,
```

Reply:
I agree it is ambiguous, here's another alternative:
```suggestion
  - if `Utxo.from_coinbase` is true, check that:
    - `spend_restriction` is `AllShieldedOutputs`, and
    - `spend_height` is greater than or equal to
      `MIN_TRANSPARENT_COINBASE_MATURITY` plus the `Utxo.height`,
    - if both checks pass, return the utxo.
    - if any check fails, drop the utxo, and let the request timeout.
```